File: Leetcode/Longest Substring Without Repeating Characters/Longest Substring Without Repeating Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        
        # last index of every character
        last_idx = {}
        max_len = 0
 
             # starting index of current 
              # window to calculate max_len
        start_idx = 0
 
        for i in range(0, len(s)):
        
        # -> Find the last index of str[i]
        # -> Update start_idx (starting index of current window)
        # -> as maximum of current value of start_idx and last
        # index plus 1
            if s[i] in last_idx:
                start_idx = max(start_idx, last_idx[s[i]] + 1)
 
                # Update result if we get a larger window
            max_len = max(max_len, i-start_idx + 1)
 
        # Update last index of current char.
            last_idx[s[i]] = i
 
        return max_len
     
''' Example 1:
Input: s = "abcabcbb"

Output: 3

Explanation: The answer is "abc", with the length of 3.
'''

File: Leetcode/Longest Substring Without Repeating Characters/Longest Substring Without Repeating Characters.java

// Given a string s, find the length of the longest substring without repeating characters.

// Example 1:
// Input: s = "abcabcbb"
// Output: 3


// Example 2:
// Input: s = "bbbbb"
// Output: 1



public int lengthOfLongestSubstring(String s) {
    // If the length of string is 0 then return 0 as length of substring for that string will also be 0
    if("".equals(s)){ 
        return 0;
    }

    // Initialize a new ArrayList
    List<Character> strings = new ArrayList<Character>();
    int longest , count  = 0;

    for(int i = 0; i < s.length() ; i++){
        Character v = s.charAt(i);
        // when repeat is found, reset back to the last occurrence prior to current.
        if(strings.contains(v)){  
            int index = s.substring(0 , i).lastIndexOf(v);
            i = index;
            count  = 0;
            strings.clear();
            continue;
        }
        // add that non repeating character to the arraylist
        strings.add(v);
        // Increment count
        count++;
        if(count > longest){
            longest = count;
        }
    }
    // return the longest substring without repeating characters
    return longest;
}